stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,561 @@
+	Fn    interpreter.ResolvedFunction `json:"fn"`
+
+	l, r *flux.TableObject
+}
+
+func (s *JoinOpSpec) IDer(ider flux.IDer) {
+	s.Left = ider.ID(s.l)
+	s.Right = ider.ID(s.r)
+}
+
+func createJoinOpSpec(args flux.Arguments, p *flux.Administration) (flux.OperationSpec, error) {
+	l, err := args.GetRequiredObject("left")
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(josh): The type system should ensure that this
+	// assertion is redundant. Unfortunately it does not.
+	// Remove this check when type inference is fixed.
+	//
+	left, ok := l.(*flux.TableObject)
+	if !ok {
+		return nil, errors.New(codes.Invalid, "argument 'left' must be a table stream")
+	}
+	p.AddParent(left)
+
+	r, err := args.GetRequiredObject("right")
+	if err != nil {
+		return nil, err
+	}
+
+	// Same comment as above. The type system should ensure
+	// that the folowing cast never panics.
+	//
+	right, ok := r.(*flux.TableObject)
+	if !ok {
+		return nil, errors.New(codes.Invalid, "argument 'right' must be a table stream")
+	}
+	p.AddParent(right)
+
+	f, err := args.GetRequiredFunction("fn")
+	if err != nil {
+		return nil, err
+	}
+
+	fn, err := interpreter.ResolveFunction(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return &JoinOpSpec{
+		Fn: fn,
+		l:  left,
+		r:  right,
+	}, nil
+}
+
+func newJoinOp() flux.OperationSpec {
+	return new(JoinOpSpec)
+}
+
+func (s *JoinOpSpec) Kind() flux.OperationKind {
+	return joinKind
+}
+
+type MergeJoinProcedureSpec struct {
+	plan.DefaultCost
+
+	Fn interpreter.ResolvedFunction `json:"fn"`
+}
+
+func newMergeJoinProcedure(spec flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	s, ok := spec.(*JoinOpSpec)
+	if !ok {
+		return nil, errors.Newf(codes.Internal, "invalid spec type %T", spec)
+	}
+	return &MergeJoinProcedureSpec{Fn: s.Fn}, nil
+}
+
+func (s *MergeJoinProcedureSpec) Kind() plan.ProcedureKind {
+	return joinKind
+}
+func (s *MergeJoinProcedureSpec) Copy() plan.ProcedureSpec {
+	return &MergeJoinProcedureSpec{Fn: s.Fn.Copy()}
+}
+
+func createMergeJoinTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	s, ok := spec.(*MergeJoinProcedureSpec)
+	if !ok {
+		return nil, nil, errors.Newf(codes.Internal, "invalid spec type %T", spec)
+	}
+	parents := a.Parents()
+
+	c := NewMergeJoinCache(a.Context(), a.Allocator(), s.Fn, parents[0], parents[1])
+	d := execute.NewDataset(id, mode, c)
+	t := NewMergeJoinTransformation(d, c)
+	return t, d, nil
+}
+
+type mergeJoinTransformation struct {
+	mu    sync.Mutex
+	d     execute.Dataset
+	cache *mergeJoinCache
+	done  bool
+}
+
+func NewMergeJoinTransformation(d execute.Dataset, cache *mergeJoinCache) *mergeJoinTransformation {
+	return &mergeJoinTransformation{
+		d:     d,
+		cache: cache,
+	}
+}
+
+func (t *mergeJoinTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	t.mu.Lock()
+	t.cache.insert(id, tbl)
```

Comment:
This won't work. You must call `Do` on this table if you plan on wanting to get more tables.